src/tike/operators/numpy/usfft.py

```diff
@@ -0,0 +1,230 @@
+
+The USFFT, NUFFT, or NFFT is a fast-fourier transform from an uniform domain to
+a non-uniform domain or vice-versa. This module provides forward Fourier
+transforms for those two cased. The inverser Fourier transforms may be created
+by negating the frequencies on the non-uniform grid.
+"""
+import numpy as np
+
+
+def _get_kernel(xp, pad, mu):
+    """Return the interpolation kernel for the USFFT."""
+    xeq = xp.mgrid[-pad:pad, -pad:pad, -pad:pad]
+    return xp.exp(-mu * xp.sum(xeq**2, axis=0)).astype('float32')
+
+
+def vector_gather(xp, Fe, x, n, m, mu):
+    """A faster implementation of sequential_gather"""
+    cons = [xp.sqrt(xp.pi / mu)**3, -xp.pi**2 / mu]
+    delta = lambda l, i, x: ((l - m + i).astype('float32') / (2 * n) - x)**2
+    F = xp.zeros(x.shape[0], dtype="complex64")
+    ell = ((2 * n * x) // 1).astype(xp.int32)  # nearest grid to x
+    for i0 in range(2 * m):
+        delta0 = delta(ell[:, 0], i0, x[:, 0])
+        for i1 in range(2 * m):
+            delta1 = delta(ell[:, 1], i1, x[:, 1])
+            for i2 in range(2 * m):
+                delta2 = delta(ell[:, 2], i2, x[:, 2])
+                Fkernel = cons[0] * xp.exp(cons[1] * (delta0 + delta1 + delta2))
+                F += Fe[n + ell[:, 0] + i0, n + ell[:, 1] + i1,
+                        n + ell[:, 2] + i2] * Fkernel
+    return F
+
+
+def sequential_gather(xp, Fe, x, n, m, mu):
+    """Gather F from the regular grid.
+
+    Parameters
+    ----------
+    Fe : (n, n, n) complex64
+        The function at equally spaced frequencies.
+    x : (N, 3) float32
+        The non-uniform frequencies.
+
+    Returns
+    -------
+    F : (N, ) complex64
+        The values at the non-uniform frequencies.
+    """
+    cons = [xp.sqrt(xp.pi / mu)**3, -xp.pi**2 / mu]
+    F = xp.zeros(x.shape[0], dtype="complex64")
+    for k in range(x.shape[0]):
+        ell0 = xp.int(xp.floor(2 * n * x[k, 0]))
+        ell1 = xp.int(xp.floor(2 * n * x[k, 1]))
+        ell2 = xp.int(xp.floor(2 * n * x[k, 2]))
+        for i0 in range(2 * m):
+            for i1 in range(2 * m):
+                for i2 in range(2 * m):
+                    kera = cons[0] * xp.exp(cons[1] * (
+                        + ((ell0 - m + i0) / (2 * n) - x[k, 0])**2
+                        + ((ell1 - m + i1) / (2 * n) - x[k, 1])**2
+                        + ((ell2 - m + i2) / (2 * n) - x[k, 2])**2
+                    ))  # yapf: disable
+                    F[k] += Fe[n + ell0 + i0, n + ell1 + i1,
+                               n + ell2 + i2] * kera
+    return F
+
+
+def eq2us(f, x, n, eps, xp, gather=vector_gather):
+    """USFFT from equally-spaced grid to unequally-spaced grid.
+
+    Parameters
+    ----------
+    f : (n, n, n) complex64
+        The function to be transformed on a regular-grid of size n.
+    x : (N, 3)
+        The sampled frequencies on unequally-spaced grid.
+    eps : float
+        The desired relative?/absolute? accuracy of the USFFT.
```

Comment:
@nikitinvv, what kind of accuracy does `eps` signify?

Reply:
Formally it is 
$\|\mathcal{F}(f)-F\|_1\le \|f\|_1(2\epsilon+o(\epsilon))$, where \mathcal{F}(f) is the exact evaluation of the Fourier transform, and F is the approximation. Norms $\|\|_1$ and $\|\|_2$ are equivalent in finite dimensional spaces. So it is fine to write 'relative'.